src/cargo/ops/registry.rs

```diff
@@ -374,6 +382,61 @@ fn transmit(
     Ok(())
 }
 
+fn wait_for_publish(
+    config: &Config,
+    registry_src: SourceId,
+    pkg: &Package,
+    timeout: std::time::Duration,
+) -> CargoResult<()> {
+    let version_req = format!("={}", pkg.version());
+    let query = Dependency::parse(pkg.name(), Some(&version_req), registry_src)?;
+
+    let now = std::time::Instant::now();
+    let sleep_time = std::time::Duration::from_secs(1);
+    let mut logged = false;
+    loop {
+        let registry_description;
+        {
+            let mut registry = PackageRegistry::new(config)?;
+            let _lock = config.acquire_package_cache_lock()?;
+            registry.lock_patches();
+            registry_description = registry.describe_source(registry_src);
+
+            let summaries = loop {
+                // Exact to avoid returning all for path/git
+                match registry.query_vec(&query, QueryKind::Exact) {
+                    std::task::Poll::Ready(res) => {
+                        break res?;
+                    }
+                    std::task::Poll::Pending => registry.block_until_ready()?,
+                }
+            };
+            if !summaries.is_empty() {
+                break;
+            }
+        }
+
+        if timeout < now.elapsed() {
+            config.shell().warn(format!(
+                "publishing {} took longer than expected",
+                pkg.name(),
+            ))?;
+            break;
+        }
```

Comment:
As the current behavior doesn't error on the publish taking too long, I figured that'd be the safest behavior to use for now.